### DISCUSSION TITLE
Is it possible to generate suggestions in the middle of lines?

### DISCUSSION THREAD

**Discussion by andersodt** [community/community#7109] [Copilot Conversations]
I'm using Jetbrains RubyMine and I'm often working inside of brackets, or when I'm editing HTML, inside of tags.

Is it possible to generate suggestions while my cursor is in the middle of text on a line? For instance, if I have something like this:

```
<td data-target="<%= cursor_here %>">
```

**Comment by fcFn** [comment]
It appears that it's not possible to use completions midline: https://www.tabnine.com/tabnine-vs-github-copilot. A workaround is to insert a new line.

**Reply by tmlewin** [reply]
@fcFn how does copilot compare to tabnine pro in terms of accuracy and robustness of the code generation and autocompletion. I have never used tabnine before. Copilot can be very smart at times in code completion especially when working in a new languiage or framework that u might not be so familiar with. saves lots of stack overflow time.

**Reply by fcFn** [reply]
Sorry, @tmlewin, I have not tried the Tabnine Pro. But I find Copilot more comfortable to use since it doesn't get in the way of the IDE static analysis autocomplete suggestions unlike Tabnine.

**Reply by tmlewin** [reply]
@fcFn Thanks for the feedback. I am just curious in terms of using copilot in a work environment. Considering the fact it is a now a commercial application. How does it deal with privacy in terms of not sharing your code with github and also preventing public code from showing up as a suggestion. It is very important to keep private code from being shared, does copilot have filters to deal with that?

**Reply by fcFn** [reply]
@tmlewin Well, there is a [setting](https://docs.github.com/en/copilot/configuring-github-copilot/configuring-github-copilot-settings-on-githubcom) that prevents Copilot from using your code (however, in its comparison article I linked to previously, Tabnine claims that only it can provide complete privacy). 

Likewise, there is a Copilot setting (see the linked docs page) that should hide suggestions matching code publicly available on GitHub. On the other hand, Tabnine can be trained on your own code or your own dataset so that you can be sure that it will never suggest any publicly available code, and they also claim that their model is trained only on code with permissive licenses. 

All in all, you probably should consult a lawyer specializing in IP and copyright law before using either Copilot or Tabnine. Better safe than sorry after all.

**Comment by cmp-nct** [comment]
That is one of the needed features.
Copilot needs to see the line it is in and provide code that will extend it from that position.

Currently your only option is to remove everything after your cursor pos.
